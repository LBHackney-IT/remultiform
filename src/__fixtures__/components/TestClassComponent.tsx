import PropTypes, { ValidationMap } from "prop-types";
import React from "react";

export interface TestClassComponentProps {
  content: string;
}

class TestClassComponent extends React.Component<TestClassComponentProps> {
  static propTypes: ValidationMap<TestClassComponentProps> = {
    content: PropTypes.string.isRequired
  };

  render(): JSX.Element {
    const { content } = this.props;

    return <div>{content}</div>;
  }
}

export default TestClassComponent;
