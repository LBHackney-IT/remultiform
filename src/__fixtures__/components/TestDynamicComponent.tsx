import PropTypes from "prop-types";
import React from "react";

import {
  DynamicComponent,
  DynamicComponentControlledProps,
} from "../../component-wrapper/DynamicComponent";

export type TestDynamicComponentProps = DynamicComponentControlledProps<
  string
> & {
  content: string;
};

export const TestDynamicComponent: React.FunctionComponent<TestDynamicComponentProps> = (
  props: TestDynamicComponentProps
): JSX.Element => {
  const { content, value, onValueChange, required, disabled } = props;

  return (
    <div>
      <div>{content}</div>
      <input
        data-testid="input"
        value={value}
        onChange={(event: React.ChangeEvent<HTMLInputElement>): void => {
          onValueChange(event.target.value);
        }}
        required={required}
        disabled={disabled}
      />
    </div>
  );
};

TestDynamicComponent.displayName = "TestDynamicComponent";

TestDynamicComponent.propTypes = {
  content: PropTypes.string.isRequired,
  ...DynamicComponent.controlledPropTypes(PropTypes.string.isRequired),
};
