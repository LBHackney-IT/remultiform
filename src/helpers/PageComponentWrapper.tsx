import PropTypes from "prop-types";
import React from "react";

/**
 * A component along with the associated properties needed to render it as part
 * of a {@link Step}.
 *
 * ```ts
 * const component: PageComponent<typeof MyInput> = {
 *   key: "my-input",
 *   Component: MyInput,
 *   props: {
 *     defaultValue: "Enter something?"
 *   }
 * }
 * ```
 */
export interface PageComponent<
  ComponentType extends React.ElementType<Props>,
  Props = ComponentType extends React.ElementType<infer T> ? T : {}
> {
  /**
   * A unique identifier for this component on the page.
   */
  key: React.Key;

  /**
   * The component class, function, or tag name to render.
   */
  Component: ComponentType;

  /**
   * The props to pass to {@link PageComponent.Component}.
   */
  props: JSX.LibraryManagedAttributes<ComponentType, Props>;
}

/**
 * A wrapper for {@link PageComponent} for a {@link Step}.
 *
 * Create these using {@link wrapPageComponent}.
 */
export interface PageComponentWrapper {
  key: React.Key;

  /**
   * The function to render the component.
   *
   * @param key - The key to give the generated component. Defaults to
   * {@link PageComponentWrapper.key}.
   */
  render(key?: React.Key): JSX.Element;
}

/**
 * The proptypes of a {@link PageComponentWrapper}.
 */
export const pageComponentWrapperPropType = PropTypes.exact({
  key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  render: PropTypes.func.isRequired
}).isRequired;

/**
 * Wrap a {@link PageComponent} in a {@link PageComponentWrapper} ready to be
 * included in a {@link Step}.
 */
// This method mostly exists so we can have strict type inference on our
// components. By wrapping the component in a generic function, the type of the
// component has to be infered in order to call the function, meaning we get
// strong types with little effort.
export const wrapPageComponent = <
  ComponentType extends React.ElementType<Props>,
  Props = ComponentType extends React.ElementType<infer T> ? T : {}
>(
  component: PageComponent<ComponentType, Props>
): PageComponentWrapper => {
  const { key, Component, props } = component;

  return {
    key,
    render(k: React.Key = key): JSX.Element {
      return <Component key={k} {...props} />;
    }
  };
};
