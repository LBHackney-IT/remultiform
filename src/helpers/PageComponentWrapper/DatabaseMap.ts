import PropTypes from "prop-types";

import { NamedSchema, Schema, StoreKey, StoreNames } from "../../store/types";

/**
 * The options for {@link DatabaseMap}.
 */
export interface DatabaseMapOptions<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * The name of the {@link Database} store to fetch and update.
   */
  storeName: StoreName;

  /**
   * The key in {@link DatabaseMapOptions.storeName} to fetch and update.
   */
  key: StoreKey<DBSchema["schema"], StoreName>;
}

/**
 * A class to tell {@link WrappedPageComponent} how its
 * {@link DynamicPageComponentControlledProps} maps to the {@link Database}.
 *
 * Unlike most generic types in this library, you will need to explicitly set
 * the type parameters on construction, as there's no way to infer a
 * {@link NamedSchema} from any of the arguments to its constructor, but they
 * constrain the possible values of its properties.
 *
 * @typeparam DBSchema - The schema of the {@link Database} this maps to.
 *
 * @typeparam StoreName - The name of the {@link Database} store this maps to.
 */
export class DatabaseMap<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * The proptype validator for a {@link DatabaseMap}.
   */
  static propType: PropTypes.Requireable<
    DatabaseMap<NamedSchema<string, number, Schema>, StoreNames<Schema>>
  > = PropTypes.exact({
    storeName: (PropTypes.string as PropTypes.Requireable<StoreNames<Schema>>)
      .isRequired,
    key: PropTypes.oneOfType([
      PropTypes.string.isRequired,
      PropTypes.number.isRequired
    ]).isRequired
  });

  readonly storeName: StoreName;
  readonly key: StoreKey<DBSchema["schema"], StoreName>;

  constructor(options: DatabaseMapOptions<DBSchema, StoreName>) {
    const { storeName, key } = options;

    this.storeName = storeName;
    this.key = key;
  }
}
