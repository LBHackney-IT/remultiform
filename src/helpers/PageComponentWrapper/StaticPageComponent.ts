import PropTypes from "prop-types";

/**
 * The options for {@link StaticPageComponent}.
 */
export interface StaticPageComponentOptions<
  ComponentType extends React.ElementType<Props>,
  Props
> {
  /**
   * A unique identifier for this component on the page.
   */
  key: React.Key;

  /**
   * The component class, function, or tag name to render.
   */
  Component: ComponentType;

  /**
   * The props to pass to {@link StaticPageComponentOptions.Component}.
   */
  props: Props;
}

/**
 * A component along with the associated properties needed to render it as part
 * of a {@link Step}.
 *
 * If you need your component to interact with a {@link Database}, use
 * a {@link DynamicPageComponent} instead.
 *
 * ```ts
 * const myInput = new PageComponent({
 *   key: "my-input",
 *   Component: MyInput,
 *   props: {
 *     className: "my-input-class"
 *   }
 * });
 * ```
 */
export class StaticPageComponent<
  ComponentType extends React.ElementType<Props>,
  Props
> {
  /**
   * The proptype validator for a {@link StaticPageComponent}.
   */
  static propType: PropTypes.Requireable<
    StaticPageComponent<
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      React.ElementType<any>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      any
    >
  > = PropTypes.exact({
    key: PropTypes.oneOfType([
      PropTypes.string.isRequired,
      PropTypes.number.isRequired
    ]).isRequired,
    Component: (PropTypes.elementType as PropTypes.Requireable<
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      React.ElementType<any>
    >).isRequired,
    props: PropTypes.object.isRequired,
    defaultValue: PropTypes.any.isRequired
  });

  readonly key: React.Key;
  readonly Component: ComponentType;
  readonly props: Props;

  constructor(options: StaticPageComponentOptions<ComponentType, Props>) {
    const { key, Component, props } = options;

    this.key = key;
    this.Component = Component;
    this.props = props;
  }
}
