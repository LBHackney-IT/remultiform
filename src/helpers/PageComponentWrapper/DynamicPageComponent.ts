import PropTypes from "prop-types";

import { NamedSchema, Schema, StoreNames, StoreValue } from "../../store/types";

import { DatabaseMap } from "./DatabaseMap";

/**
 * The proptypes of the extra props injected by {@link WrappedPageComponent}
 * for connecting a {@link DynamicPageComponent} with a {@link Database}.
 *
 * When implementing a {@link DynamicPageComponent} you should make sure to
 * hook these props up to the relevant props in the component you're wrapping in
 * order to integrate with the {@link Database} correctly.
 *
 * ```ts
 * type MyProps = DynamicPageComponentControlledProps<string> & {
 *   content: string;
 * };
 * ```
 */
export interface DynamicPageComponentControlledProps<Value> {
  /**
   * The current value of the component.
   *
   * This should follow the standard
   * {@link https://reactjs.org/docs/forms.html#controlled-components|React
   * controlled component} pattern with
   * {@link DynamicPageComponentControlledProps.onValueChange}.
   */
  value: "" | Value | null | undefined;

  /**
   * The callback to call when the component's value changes.
   *
   * This should follow the standard
   * {@link https://reactjs.org/docs/forms.html#controlled-components|React
   * controlled component} pattern with
   * {@link DynamicPageComponentControlledProps.value}.
   */
  onValueChange(value: "" | Value): void;

  /**
   * A flag for whether or not the component is disabled for input.
   *
   * {@link WrappedPageComponent} sets this to `true` when waiting for
   * {@link Database} operations to settle. The wrapped component should
   * disable functionality appropriately. If you choose not to handle this prop,
   * be aware that unsaved user entered data may be overwritten when the
   * {@link Database} operation does eventually settle.
   */
  disabled: boolean;
}

/**
 * A React component that handles {@link DynamicPageComponentControlledProps}
 * in addition to its own props.
 *
 * @typeparam Props - The proptypes for the component, excluding the
 * {@link DynamicPageComponentControlledProps}.
 *
 * @typeparam Value - The {@link StoreValue} of the value this component
 * represents in the {@link Database}.
 */
export type DynamicPageComponentType<Props, Value> = React.ComponentType<
  Props & DynamicPageComponentControlledProps<Value>
>;

/**
 * The options for {@link DynamicPageComponent}.
 */
export interface DynamicPageComponentOptions<
  ComponentType extends DynamicPageComponentType<
    Props,
    StoreValue<DBSchema["schema"], StoreName>
  >,
  Props,
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * A unique identifier for this component on the page.
   */
  key: string;

  /**
   * The component class or function to render.
   *
   * The component must implement {@link DynamicPageComponentControlledProps}
   * in addition to any of its own props.
   *
   * Pass static prop values via {@link DynamicPageComponentOptions.props}.
   */
  Component: ComponentType;

  /**
   * The static props to pass to {@link DynamicPageComponentOptions.Component}.
   *
   * These should not include the {@link DynamicPageComponentControlledProps},
   * which are injected by {@link WrappedPageComponent}.
   */
  props: Props;

  /**
   * The configuration telling the {@link DynamicPageComponent} how to interface
   * with the {@link Database}.
   *
   * See {@link DatabaseMap} for more details.
   */
  databaseMap: DatabaseMap<DBSchema, StoreName>;

  /**
   * The value to default to when nothing exists in the provided database store.
   *
   * If you leave this unspecified, it will use whatever the default value for
   * the component is.
   */
  defaultValue?: StoreValue<DBSchema["schema"], StoreName> | null;

  /**
   * The value representing an empty value for this component.
   *
   * If you leave this unspecified, it will assume the empty value is an empty
   * string.
   */
  emptyValue?: StoreValue<DBSchema["schema"], StoreName>;
}

/**
 * A component along with the associated properties needed to render it as part
 * of a {@link Step}.
 *
 * Unlike {@link StaticPageComponent}, we inject some additional
 * {@link DynamicPageComponentControlledProps} into the component to support
 * interfacing with a {@link Database}. The
 * {@link DynamicPageComponentOptions.Component} needs to support those extra
 * props if it needs to interface with a {@link Database}. If not, use a
 * {@link StaticPageComponent} instead.
 *
 * ```ts
 * const storeName = "myStore";
 *
 * const myDatabaseMap = new DatabaseMap<MyDBSchema, typeof storeName>({
 *   storeName,
 *   key: "input-0"
 * });
 *
 * const myInput = new DynamicPageComponent({
 *   key: "my-input",
 *   Component: MyInput,
 *   props: {
 *     className: "my-input-class"
 *   },
 *   defaultValue: "Nothing here...",
 *   emptyValue: "",
 *   databaseMap: myDatabaseMap
 * });
 * ```
 */
export class DynamicPageComponent<
  ComponentType extends DynamicPageComponentType<
    Props,
    StoreValue<DBSchema["schema"], StoreName>
  >,
  Props,
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * The proptype validator for a {@link DynamicPageComponent}.
   */
  static propType: PropTypes.Requireable<
    DynamicPageComponent<
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      DynamicPageComponentType<any, StoreValue<Schema, StoreNames<Schema>>>,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      any,
      NamedSchema<string, number, Schema>,
      StoreNames<Schema>
    >
  > = PropTypes.exact({
    key: PropTypes.string.isRequired,
    Component: PropTypes.func.isRequired,
    props: PropTypes.object.isRequired,
    databaseMap: DatabaseMap.propType.isRequired,
    defaultValue: PropTypes.any,
    emptyValue: PropTypes.any.isRequired
  });

  /**
   * The proptypes for {@link DynamicPageComponentControlledProps}.
   *
   * This is generic to support being specific about the type of
   * {@link DynamicPageComponentControlledProps.value}. The type should be
   * infered from the arguments.
   *
   * @param valuePropType - An optional proptype validator for the type of
   * {@link DynamicPageComponentControlledProps.value}.
   */
  static controlledPropTypes<Value>(
    valuePropType: PropTypes.Requireable<Value>
  ): PropTypes.ValidationMap<DynamicPageComponentControlledProps<Value>> {
    return {
      value: valuePropType,
      onValueChange: PropTypes.func.isRequired,
      disabled: PropTypes.bool.isRequired
    };
  }

  readonly key: string;
  readonly Component: ComponentType;
  readonly props: Props;
  readonly databaseMap: DatabaseMap<DBSchema, StoreName>;
  readonly defaultValue?: StoreValue<DBSchema["schema"], StoreName> | null;
  readonly emptyValue: "" | StoreValue<DBSchema["schema"], StoreName>;

  constructor(
    options: DynamicPageComponentOptions<
      ComponentType,
      Props,
      DBSchema,
      StoreName
    >
  ) {
    const {
      key,
      Component,
      props,
      databaseMap,
      defaultValue,
      emptyValue
    } = options;

    this.key = key;
    this.Component = Component;
    this.props = props;
    this.databaseMap = databaseMap;
    this.defaultValue = defaultValue;

    // We need to do this to keep the component as a controlled component. The
    // controlled prop must always be defined.
    this.emptyValue = emptyValue === undefined ? "" : emptyValue;
  }
}
