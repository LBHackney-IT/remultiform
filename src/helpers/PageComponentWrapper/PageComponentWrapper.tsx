import PropTypes from "prop-types";
import React from "react";

import { WrappedPageComponent } from "../../components/WrappedPageComponent";

import { NamedSchema, Schema, StoreNames, StoreValue } from "../../store/types";

import {
  DynamicPageComponentType,
  DynamicPageComponent
} from "./DynamicPageComponent";
import { StaticPageComponent } from "./StaticPageComponent";

/**
 * A wrapper for a {@link StaticPageComponent} or {@link DynamicPageComponent}
 * for use in a {@link Step}.
 *
 * Create these using {@link PageComponentWrapper.wrapStatic} and
 * {@link PageComponentWrapper.wrapDynamic}.
 */
// This mostly exists so we can have strict type inference on our components.
// By wrapping the component in a generic function, the type of the component
// has to be infered in order to call the function. Returning this non-generic
// type from that function means we can have non-generic components depend on
// it. This gives us strong typing with little effort for the user.
export class PageComponentWrapper {
  /**
   * The proptype validator for a {@link PageComponentWrapper}.
   */
  static readonly propType: PropTypes.Requireable<
    PageComponentWrapper
  > = PropTypes.exact({
    key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    element: PropTypes.element.isRequired
  });

  /**
   * Wrap a {@link StaticPageComponent} in a {@link PageComponentWrapper} ready
   * to be included in a {@link Step}.
   *
   * You shouldn't need to provide any of the type parameters. They should be
   * infered from the {@link StaticPageComponent} passed in.
   */
  static wrapStatic<Props>(
    component: StaticPageComponent<React.ElementType<Props>, Props>
  ): PageComponentWrapper {
    const { key, Component } = component;

    let element: JSX.Element;

    if (typeof Component === "string") {
      const {
        Component: IntrinsicElement,
        props
      } = component as StaticPageComponent<keyof JSX.IntrinsicElements, {}>;

      element = <IntrinsicElement key={key} {...props} />;
    } else {
      const {
        Component: ReactComponent,
        props
      } = component as StaticPageComponent<React.ComponentType<Props>, Props>;

      element = <ReactComponent key={key} {...props} />;
    }

    return new PageComponentWrapper(key, element);
  }

  /**
   * Wrap a {@link DynamicPageComponent} in a {@link PageComponentWrapper}
   * ready to be included in a {@link Step}.
   *
   * You shouldn't need to provide any of the type parameters. They should be
   * infered from the {@link DynamicPageComponent} passed in.
   */
  static wrapDynamic<
    Props,
    DBSchema extends NamedSchema<string, number, Schema>,
    StoreName extends StoreNames<DBSchema["schema"]>
  >(
    component: DynamicPageComponent<
      DynamicPageComponentType<
        Props,
        StoreValue<DBSchema["schema"], StoreName>
      >,
      Props,
      DBSchema,
      StoreName
    >
  ): PageComponentWrapper {
    const { key } = component;

    return new PageComponentWrapper(
      key,
      (
        <WrappedPageComponent<Props, DBSchema, StoreName>
          key={key}
          component={component}
        />
      ),
      true
    );
  }

  readonly key: React.Key;

  /**
   * The element for including in a page.
   */
  readonly element: JSX.Element;

  /**
   * Do not use this directly. Use {@link PageComponentWrapper.wrapStatic} or
   * {@link PageComponentWrapper.wrapDynamic} to create a new
   * {@link PageComponentWrapper}.
   *
   * @ignore
   */
  constructor(key: React.Key, element: JSX.Element, isDynamic = false) {
    this.key = key;

    if (isDynamic && element.type.name !== WrappedPageComponent.name) {
      throw new Error(
        `Expected element to be ${WrappedPageComponent.name} but ` +
          `received ${element.type}`
      );
    }

    this.element = element;
  }
}
