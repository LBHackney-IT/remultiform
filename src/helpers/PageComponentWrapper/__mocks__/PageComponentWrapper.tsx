import React from "react";

import { Database } from "../../../store/Database";
import {
  NamedSchema,
  Schema,
  StoreNames,
  StoreValue
} from "../../../store/types";

import { DatabaseMap } from "../DatabaseMap";
import {
  DynamicPageComponentType,
  DynamicPageComponent
} from "../DynamicPageComponent";
import { StaticPageComponent } from "../StaticPageComponent";

const { PageComponentWrapper: ActualPageComponentWrapper } = jest.requireActual(
  "../PageComponentWrapper"
);

export { ActualPageComponentWrapper };

export interface PageComponentWrapperRenderProps<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  database?: Database<DBSchema>;
  onChange(value: StoreValue<DBSchema["schema"], StoreName>): void;
}

export class PageComponentWrapper<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> extends ActualPageComponentWrapper<DBSchema, StoreName> {
  static wrapStatic<Props>(
    component: StaticPageComponent<React.ElementType<Props>, Props>
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): PageComponentWrapper<any, string> {
    const { key, Component } = component;

    return new PageComponentWrapper(key, () => (
      <div key={key} data-testid={key}>
        Wrapped{" "}
        {typeof Component === "string"
          ? Component
          : Component.displayName || Component.name}
      </div>
    ));
  }

  static wrapDynamic<
    Props,
    DBSchema extends NamedSchema<string, number, Schema>,
    StoreName extends StoreNames<DBSchema["schema"]>
  >(
    component: DynamicPageComponent<
      DynamicPageComponentType<
        Props,
        StoreValue<DBSchema["schema"], StoreName>
      >,
      Props,
      DBSchema,
      StoreName
    >
  ): PageComponentWrapper<DBSchema, StoreName> {
    const { key, Component, databaseMap, defaultValue, emptyValue } = component;

    return new PageComponentWrapper(
      key,
      () => (
        <div key={key} data-testid={key}>
          Wrapped {Component.displayName || Component.name}
        </div>
      ),
      databaseMap,
      defaultValue,
      emptyValue
    );
  }

  constructor(
    key: string,
    render: (
      props: PageComponentWrapperRenderProps<DBSchema, StoreName>
    ) => JSX.Element,
    databaseMap?: DatabaseMap<DBSchema, StoreName>,
    defaultValue?: StoreValue<DBSchema["schema"], StoreName> | null,
    emptyValue?: "" | StoreValue<DBSchema["schema"], StoreName>
  ) {
    super(key, render, databaseMap, defaultValue, emptyValue);
  }
}
