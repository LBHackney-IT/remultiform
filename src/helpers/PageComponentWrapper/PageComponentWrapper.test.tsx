import React from "react";
import { create } from "react-test-renderer";

import { TestClassComponent } from "../../__fixtures__/components/TestClassComponent";
import { TestDynamicComponent } from "../../__fixtures__/components/TestDynamicComponent";
import { TestFunctionComponent } from "../../__fixtures__/components/TestFunctionComponent";

import { WrappedPageComponent } from "../../components/WrappedPageComponent";

import { NamedSchema } from "../../store/types";

import { DatabaseMap } from "./DatabaseMap";
import { DynamicPageComponent } from "./DynamicPageComponent";
import { PageComponentWrapper } from "./PageComponentWrapper";
import { StaticPageComponent } from "./StaticPageComponent";

jest.mock("../../store/Database");

const storeName = "testStore";

type TestSchema = NamedSchema<
  string,
  number,
  {
    [storeName]: {
      key: number;
      value: string;
    };
  }
>;

describe(".propType", () => {
  it("exists", () => {
    expect(PageComponentWrapper.propType).toBeDefined();
  });
});

describe(".wrapStatic()", () => {
  it("returns an `PageComponentWrapper`", () => {
    const componentWrapper = PageComponentWrapper.wrapStatic(
      new StaticPageComponent({
        key: "test-class",
        Component: TestClassComponent,
        props: {
          content: "test class content"
        }
      })
    );

    expect(componentWrapper).toBeInstanceOf(PageComponentWrapper);
  });
});

describe(".wrapDynamic()", () => {
  it("returns an `PageComponentWrapper`", () => {
    const componentWrapper = PageComponentWrapper.wrapDynamic(
      new DynamicPageComponent({
        key: "test-component",
        Component: TestDynamicComponent,
        props: {
          content: "test content"
        },
        defaultValue: "test default",
        databaseMap: new DatabaseMap<TestSchema, typeof storeName>({
          storeName,
          key: 0
        })
      })
    );

    expect(componentWrapper).toBeInstanceOf(PageComponentWrapper);
  });
});

describe("new ()", () => {
  it("returns a `PageComponentWrapper` when called with any element while expecting the element to be static", () => {
    const wrapper = new PageComponentWrapper(
      "test-key",
      (<div>Not a WrappedPageComponent</div>)
    );

    expect(wrapper).toBeInstanceOf(PageComponentWrapper);
  });

  it("returns a `PageComponentWrapper` when called with a `WrappedPageComponent` element while expecting the element to be dynamic", () => {
    const key = "test-key";

    const wrapper = new PageComponentWrapper(
      key,
      (
        <WrappedPageComponent
          component={
            new DynamicPageComponent({
              key,
              Component: TestDynamicComponent,
              props: {
                content: "test content"
              },
              defaultValue: "test default",
              databaseMap: new DatabaseMap<TestSchema, typeof storeName>({
                storeName,
                key: 0
              })
            })
          }
        />
      ),
      true
    );

    expect(wrapper).toBeInstanceOf(PageComponentWrapper);
  });

  it("throws when called with a non-`WrappedPageComponent` element while expecting the element to be dynamic", () => {
    expect(
      () =>
        new PageComponentWrapper(
          "test-key",
          (<div>Not a WrappedPageComponent</div>),
          true
        )
    ).toThrowError(
      "Expected element to be WrappedPageComponent but received div"
    );
  });
});

describe("#key", () => {
  it("matches the key provided when wrapping a `PageComponent`", () => {
    const key = "test-key";

    const componentWrapper = PageComponentWrapper.wrapStatic(
      new StaticPageComponent({
        key,
        Component: TestClassComponent,
        props: {
          content: "test class content"
        }
      })
    );

    expect(componentWrapper.key).toEqual(key);
  });

  it("matches the key provided when wrapping a `DynamicPageComponent`", () => {
    const key = "test-key";

    const componentWrapper = PageComponentWrapper.wrapDynamic(
      new DynamicPageComponent({
        key,
        Component: TestDynamicComponent,
        props: {
          content: "test content"
        },
        defaultValue: "test default",
        databaseMap: new DatabaseMap<TestSchema, typeof storeName>({
          storeName,
          key: 0
        })
      })
    );

    expect(componentWrapper.key).toEqual(key);
  });
});

describe("#element", () => {
  it("renders correctly for intrinsic elements", () => {
    const componentWrapper = PageComponentWrapper.wrapStatic(
      new StaticPageComponent({
        key: "test-function",
        Component: "img",
        props: {
          src: "test.png"
        }
      })
    );

    const component = create(componentWrapper.element);

    expect(component).toMatchSnapshot();
  });

  it("renders correctly for class components", () => {
    const componentWrapper = PageComponentWrapper.wrapStatic(
      new StaticPageComponent({
        key: "test-class",
        Component: TestClassComponent,
        props: {
          content: "test class content"
        }
      })
    );

    const component = create(componentWrapper.element);

    expect(component).toMatchSnapshot();
  });

  it("renders correctly for function components", () => {
    const componentWrapper = PageComponentWrapper.wrapStatic(
      new StaticPageComponent({
        key: "test-function",
        Component: TestFunctionComponent,
        props: {
          content: "test function content"
        }
      })
    );

    const component = create(componentWrapper.element);

    expect(component).toMatchSnapshot();
  });

  it("renders correctly for components with controlled props", () => {
    const componentWrapper = PageComponentWrapper.wrapDynamic(
      new DynamicPageComponent({
        key: "test-component",
        Component: TestDynamicComponent,
        props: {
          content: "test content"
        },
        defaultValue: "test default",
        databaseMap: new DatabaseMap<TestSchema, typeof storeName>({
          storeName,
          key: 0
        })
      })
    );

    const component = create(componentWrapper.element);

    expect(component).toMatchSnapshot();
  });
});
