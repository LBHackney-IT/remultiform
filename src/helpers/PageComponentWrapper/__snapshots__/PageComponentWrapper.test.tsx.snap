// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#element renders correctly for class components 1`] = `
<div>
  test class content
</div>
`;

exports[`#element renders correctly for components with controlled props 1`] = `
<div>
  <div>
    test content
  </div>
  <input
    data-testid="input"
    disabled={true}
    onChange={[Function]}
  />
</div>
`;

exports[`#element renders correctly for function components 1`] = `
<div>
  test function content
</div>
`;

exports[`#element renders correctly for intrinsic elements 1`] = `
<img
  src="test.png"
/>
`;
